Bugfix/fix multiple connections issue

ykhatal opened the pull request:
@jkernech @PuKoren

ykhatal committed:
Use defered promise instead of connections pool array

ykhatal committed:
Change log wording

ykhatal committed:
Add prefetch examle

ykhatal committed:
Use retry module

ykhatal committed:
Add module to dependencies

ykhatal committed:
Refactor the producer :
 - remove reconnect, gracefullReconnect function
 - update produce logic

ykhatal committed:
Add timeout to default config

Commit:
Fix lint issue and bad usage of connection promise in channel resolving

Commit:
Remove connect from tests

Commit:
Refactor calls, methods, local var usage, catch

Commit:
Fix lint issue and add Logger global var to jshintrc

Commit:
Add a timer on messages to not overflow CPU when connection is lost

ykhatal committed:
Use es6 syntax

ykhatal committed:
Use es6 syntax

ykhatal committed:
Fix multiple connections

ykhatal committed:
Use es6 syntax

ykhatal committed:
Update example

mrister commented on examples/prefetch/producer.js:
```diff
@@ -0,0 +1,33 @@
+  }
+}, 500);
+
+// producer.connect()
```
@ky23  Hey are the commented out sections indicative of something (example) or they can be removed maybe ?

ykhatal replied:
@mrister i forgot to remove it, thks

ykhatal committed:
Update README.md file

mrister commented on lib/consumer.js:
```diff
@@ -5,82 +7,77 @@ var amqp = require('amqplib'),
-          amqpConnPool.shift()(amqpChannel);
-        }
+          amqpChannel = _channel;
+          resolve(amqpChannel);
```
should you return her or it was our intent to continue execution of code after resolve ?

PuKoren replied:
It is our intent because the function should return the `amqpChannel`, and that is done below (L77)

The resolve here is just for producers that ocured while disconnected, so they can all resolve because `amqpChannel` in these cases is a promise, and is returned too on L77.

It should be commented

mrister replied:
Ok makes sense.  For me personally it is a bit weird to see something get resolved mid processing because I always have some concern that it could be changed inadvertently after being resolved and that is why I am more used to seeing it at the end of execution (or with a return). This is why I asked :-)  Yeah a comment would be nice touch !

ykhatal committed:
Remove useless statements

mrister commented on lib/consumer.js:
```diff
@@ -5,82 +7,77 @@ var amqp = require('amqplib'),
-var checkRpc = function(msg, queue) {
+var checkRpc = (msg, queue) => {
   return function (_content) {
     if (_content !== undefined && msg.properties.replyTo) {
```
What if _content is somehow null ?  How come you only check if it undefined (maybe I am missing something)?

PuKoren replied:
Because we stated that null could be a returnable value.

In an RPC call, we could see something like this:

``` js
produce('getUser', { userId: 0}, { rpc: true } )
  .then((user) => {
    //user can be null if not found, but not undefined
  });
```

Because we stated that a service should not send back undefined because this means it should not be RPC, or the service does not makes sense, because undefined means nothing. We based this on our usage, maybe we are wrong.

It should be documented too!

mrister replied:
I wouldn't say that you are wrong if you can explain it and it makes sense :-)  I guess it is the matter of applied way of doing things (according). It would equally made sense, in you example, for it to return an error (e.g. UserNotFoundError) which you can then decide and implement how you wish to handle it in the .catch handler if that was the accord. But I guess it boils down to the same, you either can check for null in then or handle in the catch. Just a matter of accord in the project. :-)

PuKoren replied:
That's right. We don't yet handle errors on bunnymq for RPC calls. It is planned so we can directly use catch on the caller but we don't yet specified how to implement this. Maybe using headers in the response?

mrister replied:
Read a bit about the headers today, so yeah might be a good way to go :-)

PuKoren replied:
Created #22 to implement it

mrister commented on lib/consumer.js:
```diff
@@ -92,50 +89,53 @@ var checkRpc = function(msg, queue) {
-    }, { noAck: false });
 
-    return true;
+  return Promise.resolve()
```
Just maybe a short stylewise suggestion. Putting resolve on a new line (like a then) ?

PuKoren replied:
We should make a poll about this:

``` js
return Promise.resolve();
```

Or

``` js
return Promise
  .resolve();
```

@jkernech @ky23 ?

ykhatal replied:
For me if there is no function like in then statement, i prefer the first one.

mrister commented on lib/consumer.js:
```diff
@@ -92,50 +89,53 @@ var checkRpc = function(msg, queue) {
+  .then(createChannel)
+  .then(() => {
+    return amqpChannel.assertQueue(queue, options)
+    .then((_q) => {
```
what is _.q, i think some more descriptive name would be good (just for looking at the code).

PuKoren replied:
Agree, it should be the local scoped queue returned by assertQueue. Will update that.

Commit:
Rename _q to _queue
https://github.com/dial-once/node-bunnymq/pull/21#discussion_r56380796

Commit:
Add some comments
Related to https://github.com/dial-once/node-bunnymq/pull/21#discussion_r56377513

mrister commented on lib/consumer.js:
```diff
@@ -92,50 +89,53 @@ var checkRpc = function(msg, queue) {
-    });
-  }
-
   queue += (process.env.LOCAL_QUEUE) ? process.env.LOCAL_QUEUE : '';
```
maybe

``` javascript
queue += process.env.LOCAL_QUEUE || ''
```

mrister commented on lib/consumer.js:
```diff
@@ -92,50 +114,53 @@ var checkRpc = function(msg, queue) {
+
+        Promise.resolve(parsers.in(_msg))
+        .then(callback)
+        .then(checkRpc(_msg, _queue.queue))
```
Are callback and checkRpc actually doing something async ? If not they should not be in a .then

PuKoren replied:
Why not? It's just chaining, how would you do otherwise? Can you provide a simple sample based on this 2 calls?

mrister commented on lib/helpers/utils.js:
```diff
@@ -1,5 +1,6 @@
-module.exports.getValidUrl = function(_urls) {
+module.exports.getValidUrl = (_urls) => {
   for (var i = 0, l = _urls.length; i < l; ++i) {
     if (_urls[i] && typeof _urls[i] === 'string' && (_urls[i].indexOf('amqp://') === 0 || _urls[i].indexOf('amqps://') === 0)) {
```
maybe extract _urls[i] to a var to escape array lookups for the same thing.

ykhatal replied:
@mrister i agree, will fix that

ykhatal committed:
Use temporary var to avoid array item lookup

Commit:
Use or instead of ternary
https://github.com/dial-once/node-bunnymq/pull/21#discussion_r56384521

mrister commented:
Have you guys maybe thought about doing single commit pull requests (basically rebase branch on the branch you want to merge to)? It makes it a bi easier to review and keeps the history (log) cleaner.

PuKoren commented:
@mrister this branch had too much commits that was not necessary. It should have been squashed, yup!
